File: klippy/chelper/linear_pa_compose.c
/* Plan 8 Chunk 3 Task 2 — Linear pressure-advance polynomial composer.
 *
 * See linear_pa_compose.h for the math derivation and buffer layout.
 */

#include "linear_pa_compose.h"

/* Mirror trapq.h's MOVE_QUINTIC_POLY_COEFFS to keep this file standalone
 * (no need to pull the full trapq header into the compile graph). The two
 * constants are linked by buffer-layout convention; if MOVE_QUINTIC_POLY_COEFFS
 * ever changes, update LPA_NC here too. */
#define LPA_NC 15
#define LPA_AXES 4
#define LPA_E_OFFSET 3

void
linear_pa_compose(
    int n_phases,
    double *coeff_buf,
    double axis_n_x,
    double axis_n_y,
    double axis_n_z,
    double extr_r,
    double k_pa)
{
    if (n_phases <= 0 || coeff_buf == 0)
        return;
    int p, k;
    for (p = 0; p < n_phases; ++p) {
        double *phase = coeff_buf + p * LPA_NC * LPA_AXES;
        /* First pass: compute c_proj[k] = n . c[k]_xyz into a local
         * scratch array. */
        double c_proj[LPA_NC];
        for (k = 0; k < LPA_NC; ++k) {
            double cx = phase[k * LPA_AXES + 0];
            double cy = phase[k * LPA_AXES + 1];
            double cz = phase[k * LPA_AXES + 2];
            c_proj[k] = axis_n_x * cx + axis_n_y * cy + axis_n_z * cz;
        }
        /* Second pass: write .e from c_proj and its derivative.
         *   E[k]      = extr_r * c_proj[k] + k_pa * (k + 1) * c_proj[k + 1]
         *   E[NC - 1] = extr_r * c_proj[NC - 1]
         */
        for (k = 0; k < LPA_NC - 1; ++k) {
            phase[k * LPA_AXES + LPA_E_OFFSET] =
                extr_r * c_proj[k]
                + k_pa * (double)(k + 1) * c_proj[k + 1];
        }
        phase[(LPA_NC - 1) * LPA_AXES + LPA_E_OFFSET] =
            extr_r * c_proj[LPA_NC - 1];
    }
}

File: klippy/chelper/linear_pa_compose.py
"""Python wrapper around the linear pressure-advance polynomial composer
(linear_pa_compose.c).

Plan 8 Chunk 3 Task 2: bake linear pressure advance into the planner-
emitted polynomial. The composer reads the .x/.y/.z slots of a baked
4-axis coeff_buf (the output of bs_compose / fir_compose / pass-through)
and writes the .e slot in place per the math derivation in the header.
"""
from __future__ import annotations

from typing import List, Sequence, Tuple

from klippy.chelper import get_ffi


def linear_pa_compose(
    n_phases: int,
    coeff_buf: Sequence[float],
    axis_n: Tuple[float, float, float],
    extr_r: float,
    k_pa: float,
) -> List[float]:
    """Compose linear PA into the .e slot of an n_phases * 15 * 4 buffer.

    Parameters
    ----------
    n_phases : int
        Number of phases in the buffer.
    coeff_buf : sequence[float]
        Length n_phases * 15 * 4. Layout per coefficient:
        (.x, .y, .z, .e). The .x/.y/.z slots are read; .e is overwritten
        in the returned copy.
    axis_n : (n_x, n_y, n_z)
        Unit XY direction along motion (used to project XYZ position into
        a 1-D arc-length polynomial).
    extr_r : float
        Extruder ratio: filament-mm per XY-arc-mm (signed).
    k_pa : float
        Linear pressure-advance coefficient.

    Returns
    -------
    list[float]
        A length-(n_phases * 15 * 4) list with the .e slot filled. The
        .x/.y/.z slots are preserved bit-identically.
    """
    ffi, lib = get_ffi()
    expected = n_phases * 15 * 4
    if len(coeff_buf) != expected:
        raise ValueError(
            f"coeff_buf length {len(coeff_buf)} != expected {expected}"
        )
    buf = ffi.new("double[]", list(coeff_buf))
    lib.linear_pa_compose(
        int(n_phases), buf,
        float(axis_n[0]), float(axis_n[1]), float(axis_n[2]),
        float(extr_r), float(k_pa),
    )
    return [buf[i] for i in range(expected)]
